examples/tests/kv_interactive/main.c

```diff
@@ -0,0 +1,145 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <console.h>
+#include <kv_system.h>
+
+#define KEY_LEN  64
+#define DATA_LEN 64
+
+uint8_t key_buf[KEY_LEN];
+uint8_t data_buf[DATA_LEN];
+uint8_t value_buf[DATA_LEN];
+
+char read_buf[64];
+
+int get_command(void);
+int get_command(void) {
+  int idx = 0;
+  memset(read_buf, 0, 64);
+
+  // Read in characters until a new line
+  while (1) {
+    int c = getch();
+
+    if (c == RETURNCODE_FAIL) {
+      printf("\ngetch() failed!\n");
+      return 0;
+
+    } else {
+      char in = c;
+
+      if (in == 10 || in == 13) {
+        return 0;
+
+      } else {
+        // If this is a valid number record it
+        read_buf[idx] = in;
+        idx += 1;
+
+        // If our buffer is full, quit
+        if (idx >= 64) {
+          return 0;
+        }
+      }
+    }
+  }
+}
+
+int find_end(int start);
+int find_end(int start) {
+  for (int i = start; i < 64; i++) {
+    if (read_buf[i] == ' ' || read_buf[i] == '\n' || read_buf[i] == '\0') {
+      return i;
+    }
+  }
+  return 64;
+}
+
+int main(void) {
+  int ret;
+
+  printf("[KV Interactive] Text interface to KV store\n");
+  printf("\n");
+  printf("Please enter a command: get, set, or delete\n");
+  printf("   get <key>\n");
+  printf("   set <key> <value>\n");
+  printf("   delete <key>\n");
+  printf("\n");
+
+  while (1) {
+    ret = get_command();
+    if (ret < 0) {
+      printf("Error reading stdin\n");
+      exit(-1);
+    }
+
+    if (strncmp(read_buf, "get", 3) == 0) {
+      int key_start = 4;
+      int key_end   = find_end(key_start);
+      int key_len   = key_end - key_start;
+
+      memcpy(key_buf, read_buf + key_start, key_len);
+      key_buf[key_len] = '\0';
+
+      printf("Getting %s\n", key_buf);
+
+      ret = kv_system_get_sync(key_buf, key_len, data_buf, DATA_LEN);
+      if (ret < 0) {
+        printf("Could not get key: %i\n", ret);
+      } else {
+        printf("Got value: ");
+        for (int i = 0; i < DATA_LEN; i++) {
+          printf("%c", data_buf[i]);
+        }
+        printf("\n");
+      }
+
+    } else if (strncmp(read_buf, "set", 3) == 0) {
+      int key_start   = 4;
+      int key_end     = find_end(key_start);
+      int key_len     = key_end - key_start;
+      int value_start = key_end + 1;
+      int value_end   = find_end(value_start);
+      int value_len   = value_end - value_start;
+
+      memcpy(key_buf, read_buf + key_start, key_len);
```

Comment:
If key_end is 64 then won't this overflow?

Reply:
Worst case is it copies the entire buffer minus the "set " at the beginning, so it should fit.

Reply:
Isn't the worst case an overflow?

If `key_end` is 64 then we have the following:

```c
      int key_start   = 4;
      int key_end     = 64; // A valid result from find_end()
      int key_len     = 60;
      int value_start = 65;
      int value_end   = ??;
      int value_len   = ??;

     char read_buf[64];
```

This `memcpy()` will become:

```c
       memcpy(key_buf, read_buf + 4, 60);
```

which is ok,

```c
    key_buf[60] = '\0';
````

Seems like the wrong place to set the terminator though

but

```c
    memcpy(value_buf, read_buf + 65, ??);
````

Is an overflow